onnx/common/tensor.h

```diff
@@ -139,6 +258,295 @@ struct Tensor final {
+  }
+}
+
+template<typename F>
+inline void Tensor::apply_unary_function_double(F&& f)  {
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:  {
+      un_func<F, double, 1>(f, double_data_);
+      break;
+    }
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128:  {
+      un_func<F, double, 2>(f, double_data_);
+      break;
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+
+
+template<typename F>
+inline void Tensor::apply_binary_function_float(F&& f, const Tensor& a) {
+  check(a);
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:  {
+      bin_func<F, float, 1>(f, a, float_data_, a.floats());
+      break;
+    }
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX64:  {
+      bin_func<F, float, 2>(f, a, float_data_, a.floats());
+      break;
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+
+template<typename F>
+inline void Tensor::apply_binary_function_int32(F&& f, const Tensor& a) {
+  check(a);
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+    case ONNX_NAMESPACE::TensorProto_DataType_BOOL:
+    case ONNX_NAMESPACE::TensorProto_DataType_INT8:
+    case ONNX_NAMESPACE::TensorProto_DataType_INT16:
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32:
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT8:
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT16: {
+      bin_func<F, int32_t, 1>(f, a, int32_data_, a.int32s());
+      break;
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+
+template<typename F>
+inline void Tensor::apply_binary_function_int64(F&& f, const Tensor& a) {
+  check(a);
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64:  {
+      bin_func<F, int64_t, 1>(f, a, int64_data_, a.int64s());
+      break;
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+template<typename F>
+inline void Tensor::apply_binary_function_uint64(F&& f, const Tensor& a) {
+  check(a);
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT32:
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT64: {
+      bin_func<F, uint64_t, 1>(f, a, uint64_data_, a.uint64s());
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+
+template<typename F>
+inline void Tensor::apply_binary_function_double(F&& f, const Tensor& a) {
+  check(a);
+  switch(elem_type_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:  {
+      bin_func<F, double, 1>(f, a, double_data_, a.doubles());
+      break;
+    }
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128:  {
+      bin_func<F, double, 2>(f, a, double_data_, a.doubles());
+      break;
+    }
+    default:
+      throw("Operation not supported for this data type");
+  }
+}
+
+inline void Tensor::add(const Tensor& a) {
```

Comment:
put the macro right above these functions, also one good convention to avoid macros name clashing is undef right after all usage:
```
#define my_macro
  ...

my_macro(add, std::plus)
my_macro(subtract, std::minus)

#undef my_macro

```

Reply:
organized